lib/Platform/Intl/PlatformIntlApple.mm

```diff
@@ -26,6 +26,168 @@
 }
 }
 
+// https://unicode.org/reports/tr35/#Canonical_Unicode_Locale_Identifiers
+// https://tc39.es/ecma402/#sec-canonicalizeunicodelocaleid
+vm::CallResult<std::u16string> canonicalizeLocaleId(const std::u16string &localeId, vm::Runtime *runtime) {
+  llvh::Optional<bcp47_parser::ParsedLocaleIdentifier> parserOpt = bcp47_parser::parseLocaleId(localeId);
+
+  if (!parserOpt.hasValue()) {
+    // Not structurally valid language tag
+    return runtime->raiseRangeError(vm::TwineChar16("Unicode locale id ") + vm::TwineChar16(localeId.c_str()) +vm::TwineChar16(" is not structurally valid."));
+  }
+
+  std::u16string canoLocaleId;
+  canoLocaleId.reserve(localeId.length());
+
+  // Casing, variant, and attribute sorting are handled in parsing
+  bcp47_parser::ParsedLocaleIdentifier parsedId = parserOpt.getValue();
+  
+  // Append unicode_language_id
+  if (!parsedId.languageIdentifier.languageSubtag.empty()) {
+    canoLocaleId += parsedId.languageIdentifier.languageSubtag;
+  }
+  if (!parsedId.languageIdentifier.scriptSubtag.empty()) {
+    if (!canoLocaleId.empty()) {
+      canoLocaleId += u"-";
+    }
+    canoLocaleId += parsedId.languageIdentifier.scriptSubtag;
+  }
+  if (!parsedId.languageIdentifier.regionSubtag.empty()) {
+    if (!canoLocaleId.empty()) {
+      canoLocaleId += u"-";
+    }
+    canoLocaleId += parsedId.languageIdentifier.regionSubtag;
+  }
+  for (const auto &subtag : parsedId.languageIdentifier.variantSubtagList) {
+    if (!canoLocaleId.empty()) {
+      canoLocaleId += u"-";
+    }
+    canoLocaleId += subtag;
+  }
+
+  // Any extensions are in alphabetical order by their singleton (eg, en-t-xxx-u-yyy, not en-u-yyy-t-xxx).
+
+  // Sort other extensions by singleton
+  std::vector<std::u16string> otherExtensions;
+  for (auto it = parsedId.otherExtensionMap.begin(); it != parsedId.otherExtensionMap.end(); it++) {
+    std::u16string oExt;
+    oExt += it->first;
+    for (const auto &ext : it->second) {
+      oExt += u"-" + ext;
+    }
+    otherExtensions.push_back(u"-" + oExt);
```

Comment:
For consistency with the other places, perhaps the "-" should be added later, when appending to canoLocaleId?